main.py

```diff
@@ -1,22 +1,66 @@
-import win32com.client
+from dataclasses import dataclass
+import win32com.client as win32
+import sys
+
+if len(sys.argv) < 2:
+    print("Should be called with the name of the aspen file")
+    exit(1)
 
 # Connect to Aspen Plus
-aspen = win32com.client.Dispatch("Apwn.Document")
-aspen.InitFromFile(r"C:\Programming\delft\Aspen\cstr-ch4.apw")
-aspen.Visible = True
+aspen = win32.gencache.EnsureDispatch("Apwn.Document")
+aspen.InitFromArchive2(sys.argv[1])
+aspen.Visible = False
+aspen.SuppressDialogs = True  # Suppress windows dialogs
+
+@dataclass
+class SearchBlock:
+    data: list[tuple[str, str]]
+    children: list[str]
+
+search = {
+    "Hierarchy": SearchBlock([], ["Blocks"]),
+    "Compr": SearchBlock([("WNET", "Net Power")], []),
+    "MCompr": SearchBlock([("WNET", "Net Power")], []),
+    "Cyclone": SearchBlock([], []),
+    "Sep": SearchBlock([], []),
+    "HeatX": SearchBlock([], []),
+    "Dupl": SearchBlock([], []),
+    "Flash2": SearchBlock([], []),
+    "Heater": SearchBlock([], []),
+    "Mixer": SearchBlock([], []),
+    "Sep2": SearchBlock([], []),
+    "RPlug": SearchBlock([], []),
+    "Valve": SearchBlock([], []),
+    "RStoic": SearchBlock([], []),
+}
+
+
+data = {}
+
+def get_all_children(node):
+    return (node.Elements.Item(i) for i in range(node.Elements.Count))
 
+RECORD_TYPE = 6
 
-#Tree Is the main node in variable explorer
-blocks_node = aspen.Application.Tree.FindNode(r"\Data")
+blocks = list(get_all_children(aspen.Application.Tree.FindNode(r"\Data\Blocks")))
 
-#This is all user defined blocks
-print(blocks_node.Blocks.Elements.Count)
 # Loop through all blocks
-EnergyCost = 0
-for i in range(blocks_node.Blocks.Elements.Count):
-    block = blocks_node.Blocks.Elements.Item(i)
-    if hasattr(block.Output, 'WNET'):
-        print(f"Block {block.name} consumed {block.Output.WNET.value}kW")
-        EnergyCost += block.Output.WNET.value
-print(f"Total cost is {EnergyCost}Kw.")
+for block in blocks:
+    recordType = block.AttributeValue(RECORD_TYPE)
+    print(block.Name, block.Value, block.ValueType, recordType)
+
+    curr_data = {}
+
+    if s := search.get(recordType):
+        for path, name in s.data:
+            b = block.FindNode(rf"Output\{path}")
+            curr_data[name] = (b.Value, b.UnitString)
+            data[block.Name] = curr_data
+
+        for path in s.children:
+            b = block.FindNode(rf"Data\{path}")
+            blocks.extend(get_all_children(b))

```

Comment:
is this for handling the blocks that have blocks inside them (the custom blocks)? If so should this not be outside the if statement?

Reply:
No this should be inside it since it needs the second field of s for it